### PR TITLE
[MJAVADOC-721] Parse stderr output and suppress informational lines

### PR DESCRIPTION
**Pull request by sman-81**
This PR offers a solution for ticket MJAVADOC-721. Please refer to the ticket for details. Thanks.

Following this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MJAVADOC) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MJAVADOC-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MJAVADOC-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify -Prun-its` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by sman-81** [comment]
Hello maintainers,
could someone approve the workflow and give feedback to this pr?

**Comment by sman-81** [comment]
hello hello

**Comment by sman-81** [comment]
Rebased to master.

Hi @michael-o as you were the one to release the latest version of the plugin, I was wondering if you would take a look at my PR?

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java [hunk 1]
```diff
@@ -6144,6 +6171,23 @@ private boolean containsWarnings( String output )
...
+    private boolean isInfoOutput( String str )
+    {
+        return str == null
+                || str.startsWith( "Loading source file " )
+                || str.startsWith( "Generating " )
+                || str.startsWith( "Constructing Javadoc information" )
+                || str.startsWith( "Building index for " )
+                || str.startsWith( "Building tree for " )
+                || str.startsWith( "Standard Doclet version " );
+    }
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6189]: Where do you have this information from? From JDK source?

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java [hunk 2]
```diff
@@ -6081,7 +6081,25 @@ && isJavadocVMInitError( output ) )
                 msg.append( exitCode );
                 if ( StringUtils.isNotEmpty( err.getOutput() ) )
                 {
-                    msg.append( " - " ).append( err.getOutput() );
+                    // parse stderr, log informational output, add all other to exception message
+                    List<String> nonInfoLines = new ArrayList<>();
+                    for ( String str : err.getOutput().split( "\\R" ) )
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6086]: I assume that `getOutput()` contains everything as a string?

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java [hunk 3]
```diff
@@ -6081,7 +6081,25 @@ && isJavadocVMInitError( output ) )
...
+                    {
+                        if ( isInfoOutput( str ) )
+                        {
+                            getLog().debug( str );
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6090]: Does it make sense to send info to debug?

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java (hunk 1 above)
Comment (sman-81) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6189]: These lines are output of the Javadoc tool.

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java (hunk 2 above)
Comment (sman-81) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6086]: correct

The signature: `String StringStreamConsumer.getOutput()`

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java (hunk 3 above)
Comment (sman-81) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6090]: The method was named `isInfoOutput` in the sense of: is informational (e.g. not error) output of the Javadoc tool.
The mojo logs at debug and info levels in various places. It's a matter of taste whether the output identified by `isInfoOutput` should be logged at debug or info level. I am open to whichever. The naming of the method lets one expect that it should be logged as `info`, so I understand why you would ask the question.

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java (hunk 1 above)
Comment (michael-o) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6189]: That I understand, but is this list complete? Are those always in English? Have you tried `fr_FR`, `de_AT`?

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java (hunk 3 above)
Comment (michael-o) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6090]: If it is inconsistent in our code, I'd like to ask to file a JIRA issue about this not to be forgotten.

Comment (sman-81) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6090]: No, the existing code is just fine in my opinion.
The informational output of the tool is very verbose. I would therefore log at debug level only.

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java (hunk 1 above)
Comment (sman-81) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6189]: The stdout/err output of the tool appears to be always English. I have tested with a variety of different locales.
The tool has a command line option `-locale name` which affects the generated output (i.e. the html files), not the technical output of the tool itself.

Comment (sman-81) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6189]: I dug further into this. The Javadoc tool has localizations for Japanese and Chinese only it seems.
Found in  AdoptOpenJDK / openjdk-jdk11u: https://github.com/AdoptOpenJDK/openjdk-jdk11u/tree/master/src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/resources

So for those two my PR won't do anything, the plugin will behave just as before. The implementation is defensive.

**Comment by sman-81** [comment]
Good morning @michael-o,
are you willing to merge this pr or are there ToDos left for me?

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java [hunk 4]
```diff
@@ -6144,6 +6171,23 @@ private boolean containsWarnings( String output )
...
 
+    /**
+     * Determines whether the specified string is informational output of the Javadoc tool.
+     *
+     * @param str string to check
+     * @return true if informational output, false if not or cannot be determined
+     */
+    private boolean isInfoOutput( String str )
+    {
+        return str == null
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6182]: Last question. Why do you consider a null string informational? Do you expect a line to be `null` and not `""`?

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java [hunk 5]
```diff
@@ -6144,6 +6171,23 @@ private boolean containsWarnings( String output )
...
+                || str.startsWith( "Loading source file " )
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6183]: I think this is a typo. See:
```
13392  [WARNING] Javadoc Warnings
13393  [WARNING] Loading source files for package com.mycompany.myapp...
13394 -[WARNING] Constructing Javadoc information...
13395 -[WARNING] Building index for all the packages and classes...
13396 -[WARNING] Standard Doclet version 17.0.3+7-1
13397 -[WARNING] Building tree for all the packages and classes...
13398 -[WARNING] Generating /var/osipovmi/Projekte/maven-javadoc-plugin/target/it/MJAVADOC-259/target/site/apidocs/com/mycompany/myapp/App.html...
13399 +[DEBUG] Constructing Javadoc information...
13400 +[DEBUG] Building index for all the packages and classes...
13401 +[DEBUG] Standard Doclet version 17.0.3+7-1
13402 +[DEBUG] Building tree for all the packages and classes...
13403 +[DEBUG] Generating /var/osipovmi/Projekte/maven-javadoc-plugin/target/it/MJAVADOC-259/target/site/apidocs/com/mycompany/myapp/App.html...
```

am I right?

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java (hunk 4 above)
Comment (sman-81) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java:6182]: A null string should not go to the non-informational output therefore it is considered informational. The null check is also to ensure the following statements will never cause NPE.
You are right, I should include empty and blank strings.

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java (hunk 5 above)
Comment (sman-81) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java]: Thanks for your feedback @michael-o 

PR updated forced pushed.

Comment (michael-o) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java]: I would prefer the original prefixes from the resource bundles for consistency reasons.

Comment (sman-81) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java]: Ok. Valid point.

Comment (sman-81) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractJavadocMojo.java]: Code amended. Hope I got it now :)

**Review by michael-o — APPROVED** [review]
Build and compared output. Muss less noise. Nice change.

**Comment by sman-81** [comment]
> Build and compared output. Muss less noise. Nice change.

Thanks! I look forward to this feature in the next release. It was great working together @michael-o

**Comment by michael-o** [comment]
> > Build and compared output. Muss less noise. Nice change.
> 
> Thanks! I look forward to this feature in the next release. It was great working together @michael-o

Sure, I always appreciate high quality, straightforward patches.

**Comment by jira-importer** [comment]
Resolve #1091

**Pull request closed without merging**